### PR TITLE
feature/test-pipeline-gcloud

### LINKED ISSUE #198: Create Test pipeline definition and assets for Google Cloud
REQUIREMENTS

- [ ] Be compliant with Hangar way of working.

SUBTASKS
- [ ] Node
- [ ] Angular
- [ ] Quarkus
- [x] #267 
- [x] #269

### PR DESCRIPTION
**Pull request by mcerverc**
closes #198

### CODE REVIEW

**Comment by lpinon** [comment]
Add trigger for next pipeline

File: documentation/src/gcloud/setup-test-pipeline.asciidoc [hunk 1]
```diff
@@ -0,0 +1,10 @@
+:provider: Google Cloud
+:pipeline_type: pipeline
+:trigger_sentence: This pipeline will be configured in order to be triggered every time the build pipeline is executed successfully on a commit
+:pipeline_type2: pipeline
+:path_provider: gcloud
+:build-pipeline:
+:extra_args_quarkus: --build-pipeline-name quarkus-project-build
+:extra_args_node: --build-pipeline-name node-project-build
+:extra_args_angular: --build-pipeline-name angular-project-build
```
Comment (albertdb) [inline documentation/src/gcloud/setup-test-pipeline.asciidoc:9]: Missing openBrowserFlag and machineType. Check https://github.com/devonfw/hangar/pull/277/files#diff-d922906ac1017e0ed6268413f541beef53e5946be5a672b58024014fa0c41087

Will need to modify other providers docs to include first also.

**Review by albertdb — CHANGES_REQUESTED** [review]

**Review by albertdb — DISMISSED** [review]
Missing Python and Flutter examples in docs.

**Comment by lpinon** [comment]
> Missing Python and Flutter examples in docs.

Python done, Flutter out of scope for this sprint.

**Review by ultymatom — DISMISSED** [review]

File: documentation/src/azure-devops/setup-test-pipeline.asciidoc [hunk 2]
```diff
@@ -7,4 +7,6 @@
 :extra_args_quarkus: --build-pipeline-name quarkus-project-build
 :extra_args_node: --build-pipeline-name node-project-build
 :extra_args_angular: --build-pipeline-name angular-project-build
+:extra_args_python: --language-version 3.10
```
Comment (albertdb) [inline documentation/src/azure-devops/setup-test-pipeline.asciidoc:10]: Probably missing Build pipeline name.

Missing Extra args Flutter.

File: documentation/src/common_templates/setup-test-pipeline.asciidoc [hunk 3]
```diff
@@ -4,8 +4,15 @@ In this section we will create a test {pipeline_type} on {provider} for running
...
 
+ifndef::no-PR-or-MR[]
 Then, a Pull Request (PR) will be created in order to merge the new branch into the appropriate branch (provided in `-b` flag). The PR will be automatically merged if the repository policies are met. If the merge is not possible, either the PR URL will be shown as output, or it will be opened in your web browser if using `-w` flag.
 
+endif::[]
+ifdef::no-PR-or-MR[]
+Then, the new branch will be merged into the appropriate branch (provided in `-b` flag).
+
+endif::[]
+
```
Comment (albertdb) [inline documentation/src/common_templates/setup-test-pipeline.asciidoc:15]:
```suggestion
```

File: documentation/src/common_templates/setup-test-pipeline.asciidoc [hunk 4]
```diff
@@ -29,9 +36,11 @@ pipeline_generator.sh \
   -l <language or framework> \
   -d <project local path> \
 ifdef::build-pipeline[  --build-pipeline-name <build {pipeline_type} name> \]
+  [--language-version <version>]
```
Comment (albertdb) [inline documentation/src/common_templates/setup-test-pipeline.asciidoc:39]:
```suggestion
  --language-version <version> \
```

File: documentation/src/common_templates/setup-test-pipeline.asciidoc [hunk 5]
```diff
@@ -49,8 +58,10 @@ endif::[]
 -d, --local-directory       [Required] Local directory of your project.
 ifdef::build-pipeline[    --build-pipeline-name   [Required] Build {pipeline_type} name.]
 -a, --artifact-path                    Path to be persisted as an artifact after {pipeline_type} execution, e.g. where the application stores logs or any other blob on runtime.
+--language-version                     Version number of the language. Only required for python.
```
Comment (albertdb) [inline documentation/src/common_templates/setup-test-pipeline.asciidoc:61]:
```suggestion
    --language-version   [Required, if Flutter or Python] Language or framework version.
```

File: documentation/src/gcloud/setup-test-pipeline.asciidoc [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+:extra_args_python: --language-version 3.10
```
Comment (albertdb) [inline documentation/src/gcloud/setup-test-pipeline.asciidoc:10]: Probably missing Build pipeline name.

Missing Extra args Flutter.

File: documentation/src/github/setup-test-pipeline.asciidoc [hunk 7]
```diff
@@ -6,5 +6,8 @@
 :extra_args_quarkus:
 :extra_args_node:
 :extra_args_angular:
+:extra_args_python: --language-version 3.10
+:extra_args_flutter: --language-version 3.0.4
```
Comment (albertdb) [inline documentation/src/github/setup-test-pipeline.asciidoc:10]: Probably missing Build pipeline name.

**Review by lpinon — DISMISSED** [review]

**Review by ultymatom — APPROVED** [review]

File: scripts/pipelines/gcloud/templates/test/test-pipeline.cfg [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+# YAML file name.
+yamlFile="test-pipeline.yml"
+# Script name.
+scriptFile="test.sh"
+# Source branch.
+sourceBranch="feature/test-pipeline"
+# Description of the trigger
+triggerDescription="Trigger for test pipeline"
+# Branch trigger
+branchTrigger='.*'
```
Comment (albertdb) [inline scripts/pipelines/gcloud/templates/test/test-pipeline.cfg:17]: Unused?

File: scripts/pipelines/gcloud/templates/test/test-pipeline.cfg [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+function copyScript {
+    # Copy the script.
+    cp "${hangarPath}/${commonTemplatesPipelinePath}/${language}-${scriptFile}" "${localDirectory}/${scriptFilePath}/${scriptFile}"
+}
+
+# Function that adds the variables to be used in the pipeline.
+function addPipelineVariables {
+
+    # Setting the list of variable that will be used for this trigger
+    subsitutionVariable="_DOCKER_IMAGE=${dockerImage},_BUILD_PIPELINE_NAME=${buildPipelineName},_BRANCH_NAME="
```
Comment (albertdb) [inline scripts/pipelines/gcloud/templates/test/test-pipeline.cfg:38]: Branch name empty

**Review by albertdb — CHANGES_REQUESTED** [review]

**Review by ultymatom — APPROVED** [review]

File: documentation/src/azure-devops/setup-test-pipeline.asciidoc (hunk 2 above)
Comment (ultymatom) [reply documentation/src/azure-devops/setup-test-pipeline.asciidoc:10]: yes it is missing

File: scripts/pipelines/gcloud/templates/test/test-pipeline.cfg (hunk 8 above)
Comment (ultymatom) [reply scripts/pipelines/gcloud/templates/test/test-pipeline.cfg:17]: it is used in the add trigger, not part of this PR but in master

File: scripts/pipelines/gcloud/templates/test/test-pipeline.cfg (hunk 9 above)
Comment (ultymatom) [reply scripts/pipelines/gcloud/templates/test/test-pipeline.cfg:38]:
```suggestion
    subsitutionVariable="_DOCKER_IMAGE=${dockerImage},_BUILD_PIPELINE_NAME=${buildPipelineName},_BRANCH_NAME=\$BRANCH_NAME"
```

Better having value by default  that is the built-in var

File: scripts/pipelines/gcloud/templates/test/test-pipeline.cfg (hunk 8 above)
Comment (mcerverc) [reply scripts/pipelines/gcloud/templates/test/test-pipeline.cfg:17]: Used but not .* for this pipeline. I'll fix right now

Comment (mcerverc) [inline scripts/pipelines/gcloud/templates/test/test-pipeline.cfg:17]:
```suggestion
branchTrigger='manualtrigger'
```

File: scripts/pipelines/gcloud/templates/test/test-pipeline.cfg (hunk 9 above)
Comment (mcerverc) [reply scripts/pipelines/gcloud/templates/test/test-pipeline.cfg:38]: @ultymatom are you sure you can use default pipeline variables inside substitution variables?

File: scripts/pipelines/gcloud/templates/test/test-pipeline.cfg [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+# YAML file name.
+yamlFile="test-pipeline.yml"
+# Script name.
+scriptFile="test.sh"
+# Source branch.
+sourceBranch="feature/test-pipeline"
+# Description of the trigger
+triggerDescription="Trigger for test pipeline"
+# Branch trigger
+branchTrigger='manualtrigger'
```
Comment (albertdb) [inline scripts/pipelines/gcloud/templates/test/test-pipeline.cfg:17]:
```suggestion
branchTrigger='_manual_'
```
Would this work?

File: scripts/pipelines/gcloud/templates/test/test-pipeline.cfg (hunk 9 above)
Comment (ultymatom) [reply scripts/pipelines/gcloud/templates/test/test-pipeline.cfg]: yes I did it for the deploy

File: documentation/src/gcloud/setup-test-pipeline.asciidoc (hunk 6 above)
Comment (lpinon) [reply documentation/src/gcloud/setup-test-pipeline.asciidoc:10]:
```suggestion
:extra_args_python: --build-pipeline-name python-project-build --language-version 3.10
```

File: documentation/src/azure-devops/setup-test-pipeline.asciidoc (hunk 2 above)
Comment (lpinon) [reply documentation/src/azure-devops/setup-test-pipeline.asciidoc:10]:
```suggestion
:extra_args_python: --build-pipeline-name python-project-build --language-version 3.10
```

File: documentation/src/azure-devops/setup-test-pipeline.asciidoc [hunk 11]
```diff
@@ -7,4 +7,7 @@
...
+:extra_args_flutter: --language-version 3.3.4
```
Comment (albertdb) [inline documentation/src/azure-devops/setup-test-pipeline.asciidoc:11]: Missing Build pipeline name.

Comment (lpinon) [inline documentation/src/azure-devops/setup-test-pipeline.asciidoc:11]:
```suggestion
:extra_args_flutter: --build-pipeline-name flutter-project-build --language-version 3.3.4
```

File: documentation/src/gcloud/setup-test-pipeline.asciidoc [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+:extra_args_flutter: --language-version 3.3.4
```
Comment (lpinon) [inline documentation/src/gcloud/setup-test-pipeline.asciidoc:11]:
```suggestion
:extra_args_flutter: --language-version 3.3.4 --build-pipeline-name python-project-build
```

File: documentation/src/common_templates/setup-test-pipeline.asciidoc [hunk 13]
```diff
@@ -49,8 +57,10 @@ endif::[]
 -d, --local-directory       [Required] Local directory of your project.
 ifdef::build-pipeline[    --build-pipeline-name   [Required] Build {pipeline_type} name.]
 -a, --artifact-path                    Path to be persisted as an artifact after {pipeline_type} execution, e.g. where the application stores logs or any other blob on runtime.
+    --language-version   [Required, if Flutter or Python] Language or framework version.
 -b, --target-branch                    Name of the branch to which the Pull Request will target. PR is not created if the flag is not provided.
--w                                     Open the Pull Request on the web browser if it cannot be automatically merged. Requires -b flag.
+ifndef::no-PR-or-MR[-w                                  Open the Pull Request on the web browser if it cannot be automatically merged. Requires -b flag.]
+ifdef::machineType[-m, --machine-type                  Machine type for {pipeline_type} runner. Accepted values: E2_HIGHCPU_8, E2_HIGHCPU_32, N1_HIGHCPU_8, N1_HIGHCPU_32.]
```
Comment (albertdb) [inline documentation/src/common_templates/setup-test-pipeline.asciidoc:63]:
```suggestion
ifndef::no-PR-or-MR[-w                                     Open the Pull Request on the web browser if it cannot be automatically merged. Requires -b flag.]
ifdef::machineType[-m, --machine-type                     Machine type for {pipeline_type} runner. Accepted values: E2_HIGHCPU_8, E2_HIGHCPU_32, N1_HIGHCPU_8, N1_HIGHCPU_32.]
```

File: documentation/src/common_templates/setup-test-pipeline.asciidoc [hunk 14]
```diff
@@ -49,8 +57,10 @@ endif::[]
 -d, --local-directory       [Required] Local directory of your project.
 ifdef::build-pipeline[    --build-pipeline-name   [Required] Build {pipeline_type} name.]
 -a, --artifact-path                    Path to be persisted as an artifact after {pipeline_type} execution, e.g. where the application stores logs or any other blob on runtime.
+    --language-version   [Required, if Flutter or Python] Language or framework version.
```
Comment (albertdb) [inline documentation/src/common_templates/setup-test-pipeline.asciidoc:60]:
```suggestion
    --language-version      [Required, if Flutter or Python] Language or framework version.
```

**Review by albertdb — APPROVED** [review]

File: scripts/pipelines/gcloud/templates/test/test-pipeline.yml.template [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
+- id: "Unpack artifact"
+  name: ${_DOCKER_IMAGE}
+  script: |
+    #!/bin/bash
+    tar -xf build_*.tar
+
+- id: "Execute test script"
+  name: ${_DOCKER_IMAGE}
+  args: ['sh', '.pipelines/scripts/test.sh']
```
Comment (albertdb) [inline scripts/pipelines/gcloud/templates/test/test-pipeline.yml.template:21]:
```suggestion
  args: ['.pipelines/scripts/test.sh']
```

File: scripts/pipelines/common/templates/test/flutter-test.sh [hunk 16]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
```
Comment (albertdb) [inline scripts/pipelines/common/templates/test/flutter-test.sh:1]: Make file executable

File: scripts/pipelines/common/templates/test/python-test.sh [hunk 17]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
```
Comment (albertdb) [inline scripts/pipelines/common/templates/test/python-test.sh:1]: Make file executable

**Review by albertdb — CHANGES_REQUESTED** [review]

**Review by albertdb — APPROVED** [review]

**Review by albertdb — APPROVED** [review]

**Review by lpinon — APPROVED** [review]

File: scripts/pipelines/gcloud/templates/test/test-pipeline.cfg (hunk 10 above)
Comment (mcerverc) [reply scripts/pipelines/gcloud/templates/test/test-pipeline.cfg]: yeah, my fault

**Pull request merged**